magpie/register.py

```diff
@@ -712,7 +717,8 @@ def _parse_resource_path(permission_config_entry,   # type: PermissionConfigItem
                          entry_index,               # type: int
                          service_info,              # type: JSON
                          cookies_or_session=None,   # type: CookiesOrSessionType
-                         magpie_url=None,           # type: Optional[Str]
+                         magpie_url=None,           # type: Optional[Str],
+                         raise_errors=False         # type: bool
```

Comment:
Following our discussion, `raise_errors` was needed because `_log_permission` didn't have to raise each time, but now it seems it is applied everywhere? 
If that is indeed the case, I prefer we go back to the original plan to have an `error_handler` function. 
I don't like that `_log_permission` raises an unrelated HTTP error for a call that is multiple layers above and in a completely different module.

Reply:
`raise_errors` is not raised everywhere, well, the only exceptions are in another function :
https://github.com/Ouranosinc/Magpie/blob/48c2e6561afe55617c43a88561ba54199bc662da/magpie/register.py#L936-L941
https://github.com/Ouranosinc/Magpie/blob/48c2e6561afe55617c43a88561ba54199bc662da/magpie/register.py#L946-L951

Should I then still keep the `raise_errors` parameter?

Reply:
What do you think of an optional `raise_error` function callable?
By default, `raise_error=None` would be replaced by a no-op `lambda msg, perm: None`. 
If `raise_error` is provided, it is not replaced and is used as is.
Then, everywhere that needs to raise would simply have `raise_error("msg", permission_config_entry)` after the `_log_permission`. 
That would give all the flexibility to the method calling `magpie_register_permissions_from_config` to specific how/which error to raise or not.

Reply:
Just link this:
```python
_log_permission("Cannot generate resource", entry_index, detail="Service [{!s}] of type [{!s}] doesn't allows any sub-resource types. ".format(svc_name, svc_type))
raise_error("Cannot generate resource", permssion_config_entry)
continue  # in case 'raise_error' was no-op, to stop processing that perm-entry
```
Indeed that implies that we repeat the message... 
Maybe `_log_permission` (or another function) could be adjusted to instead return the formatted message, which would be passed to both functions?

Reply:
If we do the msg formatting in another function, I would just remove the `_log_permission` function since it would just be a `LOGGER.log` call anyway. It would look something like :
```
msg = _format_log_permission("Cannot generate resource", entry_index, detail="Service [{!s}] of type [{!s}] doesn't allows any sub-resource types. ".format(svc_name, svc_type))
LOGGER.log(logging.WARN, msg)
raise_error(msg)
```

Are you sure we are not overthinking this though? This solution would require more repeated line of codes : we have to call the message formatter, the logger and the raise_error explicitly each time. In our initial solution, we just had to provide a boolean value to the `_log_permission` function and that's it. We had to pass that boolean value in all the related functions, but it is a bit similar in the other solution, where we have to pass the `raise_error` function callable in required functions.

Reply:
I really don't like the fact that a function designed to log permission messages ends up raising an HTTP error. 
It is out of scope of what it is expected to do and also HTTP are involved only for API calls, which makes it out of scope for the entire register module. This is why I want an alternative where the initial call by the API is the one that provides the type of error (or if any) needs to be raised. 

I'm open to other ways to implement it though if you see an easier way, just brainstorming ideas.

Reply:
An alternative would be to keep the original `raise_errors` (bool), and if enabled, raise `RegistrationConfigurationError` instead. The API call would have to catch it and re-raise it as HTTPBadRequest. 
For consistency, `_log_permission` should be renamed `_handle_permission`, where it does the log + optional raising.

Reply:
You're right about the HTTP error, it does seem out of scope of the context of the `_log_permission`.

I think I like your last suggestion best. It combines the original solution, while rescoping things in a more coherent way, and the error logging will stay more concise (avoiding the formatting + logging + raising error calls each time.)